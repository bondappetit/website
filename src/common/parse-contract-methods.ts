import type { AbiItem } from 'web3-utils';

export type ContractMethodInput = {
  type: string;
  name: string;
};

export type ContractMethod = {
  methodName: string;
  inputs: ContractMethodInput[];
  payable: boolean;
};

export const parseContractMethods = <T extends { abi: AbiItem[] }>(
  contract?: T
) => {
  return contract?.abi.reduce<Record<string, ContractMethod>>(
    (acc, { name = '', type, stateMutability, inputs }) => {
      const isFunction = type === 'function';
      const stateNotViewAndPure =
        stateMutability !== 'view' && stateMutability !== 'pure';

      if (isFunction && stateNotViewAndPure) {
        acc[name] = {
          methodName: name,
          inputs:
            inputs?.map((input) => ({
              type: input.type,
              name: input.name
            })) ?? [],
          payable: stateMutability === 'payable'
        };
      }

      return acc;
    },
    {}
  );
};
