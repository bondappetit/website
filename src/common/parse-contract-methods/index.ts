export * from './parse-contract-methods';
