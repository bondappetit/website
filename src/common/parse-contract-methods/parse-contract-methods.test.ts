import networks from '@bondappetit/networks';

import { parseContractMethods } from './parse-contract-methods';

it('parseContractMethods should return parsed methods of contract', () => {
  const [, , contract] = Object.values(networks.development.contracts);

  const parsedContractMethods = parseContractMethods(contract);

  expect(parsedContractMethods).toMatchSnapshot();
});
