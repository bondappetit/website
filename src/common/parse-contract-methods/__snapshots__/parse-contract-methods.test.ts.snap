// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseContractMethods should return parsed methods of contract 1`] = `
Object {
  "allowTransferLock": Object {
    "inputs": Array [
      Object {
        "paramName": "account",
        "paramType": "address",
      },
    ],
    "methodName": "allowTransferLock",
    "payable": false,
  },
  "approve": Object {
    "inputs": Array [
      Object {
        "paramName": "spender",
        "paramType": "address",
      },
      Object {
        "paramName": "rawAmount",
        "paramType": "uint256",
      },
    ],
    "methodName": "approve",
    "payable": false,
  },
  "burn": Object {
    "inputs": Array [
      Object {
        "paramName": "account",
        "paramType": "address",
      },
      Object {
        "paramName": "amount",
        "paramType": "uint256",
      },
    ],
    "methodName": "burn",
    "payable": false,
  },
  "delegate": Object {
    "inputs": Array [
      Object {
        "paramName": "delegatee",
        "paramType": "address",
      },
    ],
    "methodName": "delegate",
    "payable": false,
  },
  "delegateBySig": Object {
    "inputs": Array [
      Object {
        "paramName": "delegatee",
        "paramType": "address",
      },
      Object {
        "paramName": "nonce",
        "paramType": "uint256",
      },
      Object {
        "paramName": "expiry",
        "paramType": "uint256",
      },
      Object {
        "paramName": "v",
        "paramType": "uint8",
      },
      Object {
        "paramName": "r",
        "paramType": "bytes32",
      },
      Object {
        "paramName": "s",
        "paramType": "bytes32",
      },
    ],
    "methodName": "delegateBySig",
    "payable": false,
  },
  "denyTransferLock": Object {
    "inputs": Array [
      Object {
        "paramName": "account",
        "paramType": "address",
      },
    ],
    "methodName": "denyTransferLock",
    "payable": false,
  },
  "mint": Object {
    "inputs": Array [
      Object {
        "paramName": "account",
        "paramType": "address",
      },
      Object {
        "paramName": "amount",
        "paramType": "uint256",
      },
    ],
    "methodName": "mint",
    "payable": false,
  },
  "renounceOwnership": Object {
    "inputs": Array [],
    "methodName": "renounceOwnership",
    "payable": false,
  },
  "transfer": Object {
    "inputs": Array [
      Object {
        "paramName": "dst",
        "paramType": "address",
      },
      Object {
        "paramName": "rawAmount",
        "paramType": "uint256",
      },
    ],
    "methodName": "transfer",
    "payable": false,
  },
  "transferFrom": Object {
    "inputs": Array [
      Object {
        "paramName": "src",
        "paramType": "address",
      },
      Object {
        "paramName": "dst",
        "paramType": "address",
      },
      Object {
        "paramName": "rawAmount",
        "paramType": "uint256",
      },
    ],
    "methodName": "transferFrom",
    "payable": false,
  },
  "transferLock": Object {
    "inputs": Array [
      Object {
        "paramName": "dst",
        "paramType": "address",
      },
      Object {
        "paramName": "rawAmount",
        "paramType": "uint256",
      },
      Object {
        "paramName": "date",
        "paramType": "uint256",
      },
    ],
    "methodName": "transferLock",
    "payable": false,
  },
  "transferOwnership": Object {
    "inputs": Array [
      Object {
        "paramName": "newOwner",
        "paramType": "address",
      },
    ],
    "methodName": "transferOwnership",
    "payable": false,
  },
}
`;
